runtime/src/iree/base/internal/cpu.c

```diff
@@ -10,9 +10,144 @@
 #include "iree/base/internal/cpu.h"
 
 #include "iree/base/target_platform.h"
+#include "iree/base/tracing.h"
 
 //===----------------------------------------------------------------------===//
-// iree_cpu_*
+// Platform-specific processor data queries
+//===----------------------------------------------------------------------===//
+
+#if defined(IREE_PLATFORM_ANDROID) || defined(IREE_PLATFORM_LINUX)
+
+// NOTE: not all kernel versions have all of the cap bits we need defined so as
+// a practice we always define the feature bits we need locally.
+#include <sys/auxv.h>
+
+// OR's |field_bit| into |field_value| if |hwcap_bit| is set in |hwcap_value|.
+#define IREE_SET_IF_HWCAP(hwcap_value, hwcap_bit, field_value, field_bit)   \
+  (field_value) = (field_value) | iree_all_bits_set(hwcap_value, hwcap_bit) \
+                      ? field_bit                                           \
+                      : 0
+
+#if defined(IREE_ARCH_ARM_64)
+// https://docs.kernel.org/arm64/elf_hwcaps.html
+#define IREE_HWCAP_ASIMD (1 << 1)
+#define IREE_HWCAP2_I8MM (1 << 13)
+static void iree_cpu_query_data_arch_hwcaps(uint32_t hwcap, uint32_t hwcap2,
+                                            uint64_t* out_fields) {
+  IREE_SET_IF_HWCAP(hwcap, IREE_HWCAP_ASIMD, out_fields[0],
```

Comment:
discrepancy between just `ASIMD` (another name for NEON in general) and `ASIMDDP` (which means dot-product extension)